File: shared/data/billing/revenuecat/src/androidMain/kotlin/wallapp/billing/revenuecat/RevenueCatMapper.kt
package wallapp.billing.revenuecat

import com.revenuecat.purchases.EntitlementInfo
import com.revenuecat.purchases.EntitlementInfos
import com.revenuecat.purchases.Offering
import com.revenuecat.purchases.Offerings
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.ProductType
import com.revenuecat.purchases.models.GoogleStoreProduct
import com.revenuecat.purchases.models.StoreProduct
import com.revenuecat.purchases.models.googleProduct
import wallapp.billing.purchase.BillingPurchase
import wallapp.billing.purchase.BillingPurchaseId.BillingPurchaseIdRevenueCat
import wallapp.billing.purchase.BillingPurchaseRevenueCat
import wallapp.billing.purchase.BillingPurchases
import wallapp.billing.sku.BillingProductId
import wallapp.billing.sku.BillingProductType
import wallapp.billing.sku.BillingSku
import wallapp.billing.sku.BillingSkuSpec

object RevenueCatMapper {

    val EntitlementInfo.billingProductId: BillingProductId
        get() = BillingProductId.from(productIdentifier, identifier, productPlanIdentifier)

    fun mapEntitlementInfosToBillingSkus(entitlementInfos: EntitlementInfos): BillingPurchases? {
        return mapEntitlementInfosToBillingSkus(entitlementInfos.all.values.toList())
    }

    fun mapEntitlementInfosToBillingSkus(entitlementInfos: List<EntitlementInfo>):
            BillingPurchases? {
        val purchaseMap = entitlementInfos
            .groupBy(
                keySelector = {
                    when {
                        !it.isActive -> "inactive"
                        it.verification.isVerified -> "verified"
                        else -> "unverified"
                    }
                },
                valueTransform = { mapEntitlementInfoToBillingSku(it) }
            )

        val verified = purchaseMap["verified"]?.ifEmpty { null }
        val unverified = purchaseMap["unverified"]?.ifEmpty { null }
        val inactive = purchaseMap["inactive"]?.ifEmpty { null }
        if (verified == null && unverified == null && inactive == null) {
            return null
        }
        return BillingPurchases(verified, unverified, inactive)
    }

    fun mapEntitlementInfoToBillingSku(entitlementInfo: EntitlementInfo): BillingPurchase {
        val purchaseId = mapEntitlementInfoToBillingPurchaseId(entitlementInfo)

        return BillingPurchaseRevenueCat(
            id = purchaseId,
            isPurchasePending = false,
            skuSpecs = mapEntitlementInfoToBillingSkuSpecs(entitlementInfo),
            debugString = "EntitlementInfo: ${entitlementInfo.identifier}",
        )
    }

    fun mapEntitlementInfoToBillingPurchaseId(entitlementInfo: EntitlementInfo): BillingPurchaseIdRevenueCat {
        return BillingPurchaseIdRevenueCat(
            productIdentifier = entitlementInfo.productIdentifier,
            productPlanIdentifier = entitlementInfo.productPlanIdentifier,
            originalPurchaseDateEpoch = entitlementInfo.originalPurchaseDate.time,
        )
    }

    fun mapEntitlementInfoToBillingSkuSpecs(entitlementInfo: EntitlementInfo): List<BillingSkuSpec> {
        return listOf(
            BillingSkuSpec(productId = entitlementInfo.billingProductId),
        )
    }

    fun mapOfferingsToBillingSkus(offerings: Offerings): List<BillingSku> {
        return mapOfferingsToBillingSkus(offerings.all.values.toList())
    }

    fun mapOfferingsToBillingSkus(offerings: List<Offering>): List<BillingSku> {
        return offerings
            .flatMap { mapOfferingToBillingSkus(it) }
            .distinct()
    }

    fun mapOfferingToBillingSkus(offering: Offering): List<BillingSku> {
        return offering.availablePackages.map { mapPackageToBillingSku(it) }
    }

    fun mapStoreProductsToBillingSkus(storeProducts: List<StoreProduct>): List<BillingSku> {
        return storeProducts.map { mapStoreProductToBillingSku(it) }
    }

    fun mapStoreProductToBillingSku(storeProduct: StoreProduct): BillingSku {
        return BillingSku(
            productType = mapProductType(storeProduct.type),
            productId = mapProductId(storeProduct.googleProduct!!),
            title = storeProduct.title,
            description = storeProduct.description,
            priceLocalized = storeProduct.price.formatted,
            priceNumerical = storeProduct.price.amountMicros.toFloat() / 1000000f,
            priceCurrencyCode = storeProduct.price.currencyCode,
            nativeSku = storeProduct,
        )
    }

    fun mapPackageToBillingSku(revenueCatPackage: Package) = BillingSku(
        productType = mapProductType(revenueCatPackage.product.type),
        productId = mapProductId(revenueCatPackage.product.googleProduct!!),
        title = revenueCatPackage.product.title,
        description = revenueCatPackage.product.description,
        priceLocalized = revenueCatPackage.product.price.formatted,
        priceNumerical = revenueCatPackage.product.price.amountMicros.toFloat() / 1000000f,
        priceCurrencyCode = revenueCatPackage.product.price.currencyCode,
        nativeSku = revenueCatPackage.product,
    )

    fun mapProductId(googleStoreProduct: GoogleStoreProduct): BillingProductId {
        return BillingProductId.from(
            productId = googleStoreProduct.productId,
            planId = googleStoreProduct.basePlanId,
        )
    }

    fun mapProductType(productType: ProductType): BillingProductType {
        return when (productType) {
            ProductType.INAPP -> BillingProductType.InApp
            ProductType.SUBS -> BillingProductType.Subscription
            ProductType.UNKNOWN -> throw IllegalArgumentException("Unknown/unsupported product type")
        }
    }
}

File: app/ios/wallApp/Billing/RevenueCat/RevenueCatMapper.swift
//
//  RevenueCatMapper.swift
//  WallApp
//

import WallApp
import Foundation
import RevenueCat


struct VerifiablePurchase {
    let billingPurchase: BillingPurchase
    let isVerified: Bool
    let isActive: Bool
}

final class RevenueCatMapper {
    
    static func mapStoreTransactionToBillingPurchases(
        storeTransaction: StoreTransaction,
        customerInfo: CustomerInfo,
        paymentPending: Bool
    ) -> BillingPurchases {
        let billingPurchase = mapStoreTransactionToBillingPurchase(storeTransaction: storeTransaction, customerInfo: customerInfo, paymentPending: paymentPending)
        if billingPurchase.isPurchasePending {
            return BillingPurchases(verifiedPurchases: nil, unverifiedPurchases: [billingPurchase], inactivePurchases: nil)
        } else {
            return BillingPurchases(verifiedPurchases: [billingPurchase], unverifiedPurchases: nil, inactivePurchases: nil)
        }
    }

    static func mapStoreTransactionToBillingPurchase(
        storeTransaction: StoreTransaction,
        customerInfo: CustomerInfo,
        paymentPending: Bool
    ) -> BillingPurchase {
        let productId = storeTransaction.productIdentifier
        let billingPurchaseId = BillingPurchaseId.BillingPurchaseIdRevenueCat(
            productIdentifier: productId,
            productPlanIdentifier: nil,
            originalPurchaseDateEpoch: storeTransaction.purchaseDate.timeIntervalSince1970.toInt64() * 1000
        )
        let billingProductId = BillingProductId.Companion().from(
            productId: productId,
            entitlementId: nil,
            planId: nil
        )
        let billingSkuSpec = BillingSkuSpec(productId: billingProductId)

        return BillingPurchaseRevenueCat(
            id: billingPurchaseId,
            isPurchasePending: paymentPending,
            skuSpecs: [billingSkuSpec],
            debugString: "StoreTransaction: \(storeTransaction.transactionIdentifier)"
        )
    }

    static func mapEntitlementInfosToBillingSkus(entitlementInfos: EntitlementInfos) -> BillingPurchases? {
        return mapEntitlementInfosToBillingSkus(entitlementInfos: Array(entitlementInfos.all.values))
    }

    static func mapEntitlementInfosToBillingSkus(entitlementInfos: [EntitlementInfo]) -> BillingPurchases? {
        let inactive = "inactive"
        let verified = "verified"
        let unverified = "unverified"
        let purchaseMap = Dictionary(grouping: entitlementInfos) { info in
            if !info.isActive {
                return inactive
            } else if info.verification.isVerified {
                return verified
            } else {
                return unverified
            }
        }.mapValues { infos in
            infos.map { mapEntitlementInfoToBillingSku($0) }
        }
        
        let verifiedPurchases = purchaseMap[verified]?.isEmpty == true ? nil : purchaseMap[verified]
        let unverifiedPurchases = purchaseMap[unverified]?.isEmpty == true ? nil : purchaseMap[unverified]
        let inactivePurchases = purchaseMap[inactive]?.isEmpty == true ? nil : purchaseMap[inactive]
        
        guard verifiedPurchases != nil || unverifiedPurchases != nil || inactivePurchases != nil else {
            return nil
        }
        
        return BillingPurchases(verifiedPurchases: verifiedPurchases, unverifiedPurchases: unverifiedPurchases, inactivePurchases: inactivePurchases)
    }


    static func mapEntitlementInfoToBillingSku(_ entitlementInfo: EntitlementInfo) -> BillingPurchase {
        let purchaseId = mapEntitlementInfoToBillingPurchaseId(entitlementInfo: entitlementInfo)

        return BillingPurchaseRevenueCat(
            id: purchaseId,
            isPurchasePending: false,
            skuSpecs: mapEntitlementInfoToBillingSkuSpecs(entitlementInfo: entitlementInfo),
            debugString: "EntitlementInfo: \(entitlementInfo.identifier)"
        )
    }

    static func mapEntitlementInfoToBillingPurchaseId(entitlementInfo: EntitlementInfo) -> BillingPurchaseId.BillingPurchaseIdRevenueCat {
        return BillingPurchaseId.BillingPurchaseIdRevenueCat(
            productIdentifier: entitlementInfo.productIdentifier,
            productPlanIdentifier: entitlementInfo.productPlanIdentifier,
            originalPurchaseDateEpoch: (entitlementInfo.originalPurchaseDate?.timeIntervalSince1970 ?? 0).toInt64() * 1000
        )
    }

    static func mapEntitlementInfoToBillingSkuSpecs(entitlementInfo: EntitlementInfo) -> [BillingSkuSpec] {
        let billingProductId = BillingProductId.Companion().from(
            productId: entitlementInfo.productIdentifier,
            entitlementId: entitlementInfo.identifier,
            planId: nil
        )
        return [BillingSkuSpec(productId: billingProductId)]
    }

    
    static func mapToBillingSkus(offerings: Offerings) -> RevenueCatBillingSkus.Success {
        return RevenueCatBillingSkus.Success(allBillingSkus: mapOfferingsToBillingSkus(offerings: offerings))
    }
    
    static func mapOfferingsToBillingSkus(offerings: Offerings) -> [BillingSku] {
        return mapOfferingsToBillingSkus(offerings: offerings.all.values.map { $0 })
    }

    static func mapOfferingsToBillingSkus(offerings: [Offering]) -> [BillingSku] {
        return offerings
            .flatMap { mapOfferingToBillingSkus(offering: $0) }
            .distinctUnordered()
    }

    static func mapOfferingToBillingSkus(offering: Offering) -> [BillingSku] {
        return offering.availablePackages.map { mapPackageToBillingSku(revenueCatPackage: $0) }
    }

    static func mapPackageToBillingSku(revenueCatPackage: Package) -> BillingSku {
        return BillingSku(
            productType: mapProductType(productType: revenueCatPackage.storeProduct.productType),
            productId: mapProductId(product: revenueCatPackage.storeProduct),
            title: revenueCatPackage.storeProduct.localizedTitle,
            description: revenueCatPackage.storeProduct.localizedDescription,
            priceLocalized: revenueCatPackage.storeProduct.localizedPriceString,
            priceNumerical: NSDecimalNumber(decimal: revenueCatPackage.storeProduct.price).floatValue,
            priceCurrencyCode: revenueCatPackage.storeProduct.currencyCode ?? "",
            nativeSku: revenueCatPackage.storeProduct
        )
    }

    static func mapStoreProductToBillingSku(storeProduct: StoreProduct) -> BillingSku {
        let billingProductId = BillingProductId.Companion().from(
            productId: storeProduct.productIdentifier,
            entitlementId: nil,
            planId: nil
        )

        let productType = mapProductType(productType: storeProduct.productType)

        return BillingSku(
            productType: productType,
            productId: billingProductId,
            title: storeProduct.localizedTitle,
            description: storeProduct.localizedDescription,
            priceLocalized: storeProduct.localizedPriceString,
            priceNumerical: NSDecimalNumber(decimal: storeProduct.price).floatValue,
            priceCurrencyCode: storeProduct.currencyCode ?? "",
            nativeSku: storeProduct
        )
    }

    static func mapProductId(product: StoreProduct) -> BillingProductId {
        return BillingProductId.Companion().from(
            productId: product.productIdentifier,
            entitlementId: nil,
            planId: nil
        )
    }

    static func mapProductType(productType: StoreProduct.ProductType) -> BillingProductType {
        switch productType {
        case .consumable, .nonConsumable:
            return .inApp
        case .nonRenewableSubscription, .autoRenewableSubscription:
            return .subscription
        }
    }
}
